File: lintcode/python/0060_search_insert_position.py
#!/usr/bin/python -t

# binary search solution time O(logn) space O(1)

class Solution:
    """
    @param A: an integer sorted array
    @param target: an integer to be inserted
    @return: An integer
    """
    def searchInsert(self, A, target):
        # write your code here
        n = len(A)
        if n == 0:
            return 0
        l = 0
        r = n - 1
        
        while l < r:
            mid = (l+r)/2
            if A[mid] == target:
                return mid
            elif A[mid] < target:
                l = mid + 1
            else:
                r = mid
                
        if A[l] == target:
            return l
        if A[l] < target and l == n-1:
            return l+1
        return l

File: lintcode/golang/0060_search_insert_position.go
/**
 * @param A: an integer sorted array
 * @param target: an integer to be inserted
 * @return: An integer
 */
 
//import "fmt"
func searchInsert (A []int, target int) int {
    // write your code here
    n := len(A)
    //fmt.Println(n)
    if n == 0 {
        return 0
    }
    
    left := 0
    right := n-1
    
    for left + 1 < right {
        mid := (left+right)/2
        
        if A[mid] == target {
            return mid
        } else if A[mid] > target {
            right = mid
        } else {
            left = mid
        }
    }
    
    //fmt.Println(left, right)
    //fmt.Println(A[left], A[right])
    
    if A[left] >= target {
        return left
    }
    
    if right != 0 && A[right] < target {
        return right+1
    }
    
    return right
}

